hrqb/base/task.py

```diff
@@ -125,6 +126,45 @@ def run(self) -> None:
         self.target.write(self.get_dataframe())
 
 
+class SQLQueryExtractTask(PandasPickleTask):
+    """Base class for Tasks that make SQL queries for data."""
+
+    @property
+    def dwclient(self) -> DWClient:
+        """Optional property to provide a DWClient instance."""
+        return DWClient()  # pragma: nocover
+
+    @property
+    def sql_query(self) -> str | None:
+        """SQL query from string to execute."""
+        return None
+
+    @property
+    def sql_file(self) -> str | None:
+        """SQL query loaded from file to execute."""
+        return None
+
+    @property
+    def sql_query_parameters(self) -> dict:
+        """Optional parameters to include with SQL query."""
+        return {}
+
+    def get_dataframe(self) -> pd.DataFrame:
+        """Perform SQL query and return DataFrame for required get_dataframe method."""
+        if self.sql_query:
+            query = self.sql_query
+        elif self.sql_file:
+            with open(self.sql_file) as f:
```

Comment:
Where would this SQL files live and in what cases is a SQL file preferred over a string? 🤔 

Hmm, what do you think of an update like:

```
class SQLQueryExtractTask(PandasPickleTask):
    """Base class for Tasks that make SQL queries for data."""

   ...

    @property
    def sql_query(self) -> str | None:
        """SQL query from string or loaded from file"""
        if self.sql_query:
            query = self.sql_query
        elif self.sql_file:
            with open(self.sql_file) as f:
                query = f.read()
        else:
            message = "Property sql_query or sql_file must be set."
            raise AttributeError(message)
        return query

    @property
    def sql_query_string(self) -> str | None:
        """SQL query from string to execute."""
        return None

    @property
    def sql_file(self) -> str | None:
        """SQL query loaded from file to execute."""
        return None

    @property
    def sql_query_parameters(self) -> dict:
        """Optional parameters to include with SQL query."""
        return {}

    def get_dataframe(self) -> pd.DataFrame:
        return self.dwclient.execute_query(
            self.sql_query,
            params=self.sql_query_parameters,
        )
```

Reply:
The question and refactor suggestion are well received.  In fact, given some work since this PR on actually scoping out some extract tasks, it seems pretty clear that most (if not all) SQL extract tasks will use a file.

However, being able to easily override that with an explicit string is handy for testing.  I think there is a middleground here:

- `self.sql_query` returns a string, defaulting to reading `self.sql_file`
- if neither are overridden, throws an exception that `sql_file` must be defined OR `sql_query` overridden to provide a string vs reading a file

Going to make that change and  push a commit.